File: src/wiki/schemas/reports.py
"""Typed operation result models for library-first API."""

from __future__ import annotations

from pathlib import Path
from typing import Literal

from pydantic import BaseModel, Field

from .rules import Severity

IssueSeverity = Literal["error", "warning"]


class Issue(BaseModel):
    code: str
    message: str
    path: Path | None = None
    severity: IssueSeverity = "error"


class AuditReport(BaseModel):
    ok: bool = True
    errors: list[Issue] = Field(default_factory=list)
    warnings: list[Issue] = Field(default_factory=list)

    @classmethod
    def empty(cls) -> AuditReport:
        return cls()

    def merge(self, other: AuditReport) -> AuditReport:
        return AuditReport(
            ok=self.ok and other.ok,
            errors=self.errors + other.errors,
            warnings=self.warnings + other.warnings,
        )

    def apply_strict(self) -> AuditReport:
        if not self.warnings:
            return self
        return AuditReport(
            ok=False,
            errors=self.errors + self.warnings,
            warnings=[],
        )

    def messages(self) -> tuple[list[str], list[str]]:
        return [issue.message for issue in self.errors], [issue.message for issue in self.warnings]


class LinkReport(BaseModel):
    ok: bool = True
    opportunities: int = 0
    fixes: int = 0
    changed_paths: list[Path] = Field(default_factory=list)
    remaining_broken: int = 0
    lines: list[str] = Field(default_factory=list)


class RenderReport(BaseModel):
    ok: bool = True
    updated_count: int = 0
    error_count: int = 0
    stale_files: list[str] = Field(default_factory=list)
    render_errors: list[str] = Field(default_factory=list)


class BuildOptions(BaseModel):
    output_dir: Path
    render_first: bool = False
    reload_graph: bool = False
    disk_cache: bool = False
    skip_preflight: bool = False
    verbose: bool = False


class BuildResult(BaseModel):
    ok: bool = True
    page_count: int = 0
    asset_count: int = 0
    written_paths: list[Path] = Field(default_factory=list)
    preflight: AuditReport | None = None
    error_message: str | None = None


class ExportResult(BaseModel):
    ok: bool = True
    output: str = ""
    error_message: str | None = None


class FmtReport(BaseModel):
    ok: bool = True
    stale_files: list[Path] = Field(default_factory=list)
    formatted_count: int = 0
    error_message: str | None = None
    verbose_lines: list[str] = Field(default_factory=list)


class ScaffoldResult(BaseModel):
    ok: bool = True
    config_path: Path | None = None
    written_paths: list[Path] = Field(default_factory=list)
    message: str = ""
    error_message: str | None = None


def severity_for_rule(rules: object, rule_key: str) -> Severity:
    return getattr(rules, rule_key, "warning")  # type: ignore[no-any-return]

File: src/wiki/schemas/reports.ts
/**
 * Typed operation-result models for the library-first API.
 *
 * Port of `src/wiki/schemas/reports.py`. These are the values every command
 * returns and the CLI turns into output and an exit code, so their *field
 * names* are part of the API — `report.errors`, `result.written_paths` — and
 * keep Python's spelling.
 *
 * Where pydantic supplies behaviour rather than storage, the port spells it
 * out as a method, because the behaviour is what callers depend on:
 *
 * - `AuditReport.empty()` is the starting point of every audit pass, and
 *   `merge()` is how `check` and `lint` become one report.
 * - `apply_strict()` is `--strict`: warnings become errors, and the report
 *   stops being ok. Note that it does not return a copy when there is nothing
 *   to promote — callers compare identity nowhere, but the shortcut is the
 *   Python behaviour and costs nothing to keep.
 */

import type { Path } from "../fspath.ts";
import type { OutputEntry } from "./domain.ts";

/** The two severities an issue can carry. */
export type IssueSeverity = "error" | "warning";

/** One audit finding. */
export interface Issue {
  readonly code: string;
  readonly message: string;
  readonly path?: Path | null;
  readonly severity?: IssueSeverity;
}

/** The result of an audit pass: whether it passed, and why not. */
export class AuditReport {
  readonly ok: boolean;
  readonly errors: readonly Issue[];
  readonly warnings: readonly Issue[];

  constructor(
    options: {
      readonly ok?: boolean;
      readonly errors?: readonly Issue[];
      readonly warnings?: readonly Issue[];
    } = {},
  ) {
    this.ok = options.ok ?? true;
    this.errors = options.errors ?? [];
    this.warnings = options.warnings ?? [];
  }

  /** A passing report with no issues. */
  static empty(): AuditReport {
    return new AuditReport();
  }

  /** Combine two passes into one report, in the order they ran. */
  merge(other: AuditReport): AuditReport {
    return new AuditReport({
      ok: this.ok && other.ok,
      errors: [...this.errors, ...other.errors],
      warnings: [...this.warnings, ...other.warnings],
    });
  }

  /** Promote every warning to an error, as `--strict` does. */
  applyStrict(): AuditReport {
    if (this.warnings.length === 0) return this;
    return new AuditReport({
      ok: false,
      errors: [...this.errors, ...this.warnings],
      warnings: [],
    });
  }

  /** `[error messages, warning messages]`, which is what the CLI prints. */
  messages(): [string[], string[]] {
    return [
      this.errors.map((issue) => issue.message),
      this.warnings.map((issue) => issue.message),
    ];
  }
}

/** An issue list plus the severity a rule assigned it. */
export function severityIssues(
  code: string,
  messages: readonly string[],
  severity: IssueSeverity,
): Issue[] {
  return messages.map((message) => ({ code, message, severity }));
}

/** What `link fix` did. */
export interface LinkReport {
  readonly ok: boolean;
  readonly opportunities: number;
  readonly fixes: number;
  readonly changed_paths: readonly Path[];
  readonly remaining_broken: number;
  readonly lines: readonly string[];
}

/** What `render` did. */
export interface RenderReport {
  readonly ok: boolean;
  readonly updated_count: number;
  readonly error_count: number;
  readonly stale_files: readonly string[];
  readonly render_errors: readonly string[];
}

/** Options for `build`. */
export interface BuildOptions {
  readonly output_dir: Path;
  readonly render_first?: boolean;
  readonly reload_graph?: boolean;
  readonly disk_cache?: boolean;
  readonly skip_preflight?: boolean;
  readonly verbose?: boolean;
}

/** What `build` did, including the preflight report when it ran one. */
export interface BuildResult {
  readonly ok: boolean;
  readonly page_count: number;
  readonly asset_count: number;
  readonly written_paths: readonly Path[];
  readonly preflight?: AuditReport | null;
  readonly error_message?: string | null;
}

/** What `export` produced. */
export interface ExportResult {
  readonly ok: boolean;
  readonly output: string;
  readonly error_message?: string | null;
}

/** What `fmt` did. */
export interface FmtReport {
  readonly ok: boolean;
  readonly stale_files: readonly Path[];
  readonly formatted_count: number;
  readonly error_message?: string | null;
  readonly verbose_lines: readonly string[];
}

/** What `init` wrote. */
export interface ScaffoldResult {
  readonly ok: boolean;
  readonly config_path?: Path | null;
  readonly written_paths: readonly Path[];
  readonly message: string;
  readonly error_message?: string | null;
}

/** An output entry list plus the entries a collision check needs. */
export type Manifest = readonly OutputEntry[];
